java/src/jni/h5util.c

```diff
@@ -4043,6 +4043,551 @@ Java_hdf_hdf5lib_H5_H5export_1attribute(JNIEnv *env, jclass clss, jstring file_e
+                                    ((char *)&doubleValue)[x] = ((char *)objBuf)[vlSize + x];
+                                }
+
+                                jobj = ENVPTR->CallStaticObjectMethod(ENVONLY, cDouble, doubleValueMid,
+                                                                      doubleValue);
+                                CHECK_JNI_EXCEPTION(ENVONLY, JNI_FALSE);
+                                break;
+                            }
+                        }
+                        break;
+                    }
+                    case H5T_REFERENCE: {
+                        jboolean bb;
+                        jbyte   *barray = NULL;
+
+                        jsize byteArraySize = (jsize)vlSize;
+                        if (vlSize != (size_t)byteArraySize)
+                            H5_JNI_FATAL_ERROR(ENVONLY, "translate_rbuf: overflow of byteArraySize");
+
+                        if (NULL == (jobj = ENVPTR->NewByteArray(ENVONLY, byteArraySize)))
+                            CHECK_JNI_EXCEPTION(ENVONLY, JNI_FALSE);
+
+                        PIN_BYTE_ARRAY(ENVONLY, (jbyteArray)jobj, barray, &bb,
+                                       "translate_rbuf reference: byte array not pinned");
+
+                        for (x = 0; x < vlSize; x++) {
+                            barray[x] = ((jbyte *)objBuf)[vlSize + x];
+                        }
+                        if (barray)
+                            UNPIN_BYTE_ARRAY(ENVONLY, (jbyteArray)jobj, barray, jobj ? 0 : JNI_ABORT);
+                        break;
+                    }
+                    default:
+                        H5_UNIMPLEMENTED(ENVONLY, "translate_rbuf: invalid class type");
+                        break;
+                }
+                /* Add it to the list */
+                if (jobj) {
+                    ENVPTR->CallBooleanMethod(ENVONLY, jList, arrAddMethod, jobj);
+                    CHECK_JNI_EXCEPTION(ENVONLY, JNI_TRUE);
+
+                    ENVPTR->DeleteLocalRef(ENVONLY, jobj);
+                }
+            }
+            if (!found_jList) {
+                ENVPTR->CallBooleanMethod(ENVONLY, ret_buf, arrAddMethod, jList);
+                CHECK_JNI_EXCEPTION(ENVONLY, JNI_TRUE);
+                ENVPTR->DeleteLocalRef(ENVONLY, jList);
+            }
+        }
+        H5Tclose(memb);
+    }
+    else if (type_class == H5T_ARRAY) {
+        if (!(memb = H5Tget_super(mem_type_id)))
+            H5_LIBRARY_ERROR(ENVONLY);
+        if ((vlClass = H5Tget_class(memb)) < 0)
+            H5_LIBRARY_ERROR(ENVONLY);
+        if (!(vlSize = H5Tget_size(memb)))
+            H5_LIBRARY_ERROR(ENVONLY);
+
+        // TODO
+        H5_UNIMPLEMENTED(ENVONLY, "translate_rbuf: invalid H5T_ARRAY type");
+    }
+
+done:
+
+    return (jint)status;
+}
+
+herr_t
+translate_wbuf(JNIEnv *env, jobjectArray in_buf, jlong mem_type_id, H5T_class_t type_class, jsize count,
+               jobjectArray raw_buf)
```

Comment:
For writes, this result is always going to be a byte buffer that gets passed to H5Awrite/H5Dwrite, correct? In that case, I'd not mark this as returning a jobjectArray because it's misleading. In fact, we're doing direct `HDmemcpy` on a jobjectArray which might only be working by coincidence. Probably best to mark raw_buf as a `void *` or similar.

Reply:
yeah, void * would be better.

Reply:
How about jbyte * as that is the declared type of writeBuf?

Reply:
That would probably be best if that's what matches the type for writeBuf/readBuf.